runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java

```diff
@@ -211,6 +261,20 @@ boolean canTranslate(T transform, FlinkStreamingTranslationContext context) {
         if (WriteFilesTranslation.isWindowedWrites(transform)) {
           replacement = replacement.withWindowedWrites();
         }
+
+        if (options.isAutoBalanceWriteFilesShardingEnabled()) {
+          Preconditions.checkArgument(
+              options.getParallelism() > 0,
+              "Parallelism is required to be set in FlinkPipelineOptions when isAutoBalanceWriteFilesShardingEnabled");
+
+          replacement =
+              replacement.withShardingFunction(
+                  new FlinkAutoBalancedShardKeyShardingFunction<>(
+                      options.getParallelism(),
+                      options.getMaxParallelism(),
+                      sink.getDynamicDestinations().getDestinationCoder()));
+        }
+
         return PTransformReplacement.of(
             PTransformReplacements.getSingletonMainInput(transform),
             replacement.withNumShards(numShards));
```

Comment:
I have a problem here. I found out that now, my explicit number of shards gets always override here to runner determined. 
I see that `replacement` is recreated from scratch and and must be created full. It seems that `num shards` is not explicit in payload but gets injected here and there during the transform `expand` (just guessing here)?.
It seems I can not just put sharding function to existing WriteFiles context, but needs to create fresh new one. Am I right?

I need a help here. Is there an easy way of accessing `numShards`? Do I need to change `WriteFilesPayload` and make it explicit ? @reuvenlax what approach would you suggest to resolve this?

Reply:
Never mind. I talked with @mxm and noticed that overrides have access to proto and WriteFiles transform. I got derailed by runner determined sharding code and thought that accessing stuff from proto is the only way. But it seems it is now so I should be good here